### ISSUE TITLE
On-premises hosting

### ISSUE THREAD

**Issue by lamuertepeluda** [cortexlabs/cortex#782]
Hi there,

I'd be interested in using this project but I would like to know if the generated container can run on different platforms than AWS, e.g. self-hosted (on a VM that has Docker) and locally?

Would it be possible? Or using AWS is unavoidable?

Thanks

**Comment by vishalbollu** [comment]
Hi @lamuertepeluda, Cortex currently only runs on AWS. If you are looking to test Cortex, other users in the past have spun up a tiny single worker node cluster to play with Cortex.

Out of curiosity, what functionality of Cortex are you looking to use when you run Cortex locally or on a VM? If you can explain your use-case, we may be able to improve our roadmap.

**Comment by lamuertepeluda** [comment]
Hi @vishalbollu , basically we need to run the models on premises/other VMs in order to avoid vendor lock-in and costs (until, of course, the model costs are covered by revenues or fundings). 
And also because ours is highly experimental stuff most of the times. 
I thought cortex would be a good scaffolding tool for putting our models in production without having to do it manually all the times.

**Comment by ospillinger** [comment]
We'd love to support this use case but it might take some time given that we're a small team right now and our primary focus is AWS. Regarding lock-in, supporting other cloud providers like GCP is a high priority for us. As far as cost goes, it's easy to spin cortex clusters up when you're experimenting and down when you're done. You can also take advantage of the fact that Cortex can autoscale instances down to 0 when no models are deployed. We'd be happy to help you set that up if you are interested.

**Comment by lamuertepeluda** [comment]
I see, but do you think it would be difficult to make it run on minikube, for instance?

**Comment by ospillinger** [comment]
Cortex uses functionality from EKS, EC2, CloudWatch, and S3 so it would be possible to have some of the key features working on minikube without a huge amount of effort, but full compatibility would be a lot harder which is why we haven't done it yet.

**Comment by samuel-davis** [comment]
I too would love to have offline installation capability for the framework and I am willing to implement the features and make pull requests to get it. 

**Background**: I built a machine learning framework for video processing centered around Tensorflow Serving for a private application with my company.  The solution as it stands has rather limited support for other frameworks such as PyTorch and scikit-learn. 

**Reasoning**: Having the ability to install Cortex on bare metal servers would increase the adoption rate significantly in certain communities in my opinion.  Most of the clients utilizing the application have more than enough funding to self-support hosting but are extremely evasive when it comes to utilizing AWS because of what they see as long term cost.  

For instance, they have already purchased multiple server racks with Nvidia GPUs which are currently utilizing Nvidia GRID / vGPU with the application I made.  They will not want to backtrack and waste all the money they used for purchasing those server racks/GPUs by transitioning to a cloud-only solution.  

If this is a feature you all would like to have available for Cortex please get in touch with me, we can have a conversation somewhere beyond this ticket if it is convenient. Otherwise, just point me where I am needed and I will get started this weekend. 

Looking forward to hearing from you. 
-Sam

**Comment by deliahu** [comment]
I'm closing this question since we already have a GitHub issue to track our progress on on-premise deployments: https://github.com/cortexlabs/cortex/issues/150
